Documentation.app/Contents/MacOS/FP concepts.playground/Pages/Rank-N polymorphism.xcplaygroundpage/Contents.swift

```diff
@@ -0,0 +1,155 @@
+ ```
+ func callTwice<A, B>(_ f: <T>(T) -> [T], _ a: A, _ b: B) -> ([A], [B]) {
+     (f(a), f(b)) // `f` really needs to be polymorphic, because we call it twice with inputs of different types.
+ }
+ ```
+ Notice that we included `<T>` in front of the type signature of the parameter of our function `callTwice`. With this, we want to express that the parameter `f` must be a function polymorphic over one type parameter we called `T`. This is a made up syntax, don't try to write this in your swift program because it won't compile.
+
+ If we could write such a function, we could pass our `singletonArray` function to it, along with a value of any type we want:
+ ```
+ let a = callTwice(singletonArray, 0) // == [0]
+ let b = callTwice(singletonArray, "b") // == ["b"]
+ let c = callTwice(singletonArray, [0]) // == [[0]]
+ ```
+ Notice how at each call of `callTwice` we are passing a value of different type, yet we always pass the same function `singletonArray` which is called with each type with a value of different type. With the syntax we made up, this is possible because we are able to say that the function we pass to `callTwice` must be a polymorphic function, in other words a function that can work with inputs of any type.
+
+ The reason we needed to use a made up syntax for this example is because Swift doesn't allow us to write a function that requires a polymorphic function as a parameter. This feature is called *Rank-N polymorphism* and it's supported on some languages like Haskell.
+
+ In the following sections we develop a technique to simulate Rank-N polymorphism. We start by reviewing the basics of closures. We then provide an alternative but equivalent way to encode closures that we will be able to generalise into something that approaches Rank-N polymorphism.
+
+ ## The basics of closures
+ Swift supports higher-order functions, which means that we can pass a function as an argument to another function or return a function as the result of another function. Furthermore, we can define *closures* which are functions that capture some data from its environment.
+
+ For example, the function `randomWordGenerator` below returns a random word of length `n`. As you can see `n` is a random number that is generated once each time you execute this program. `randomWordGenerator` captures the value of `n`, which means that the strings returned by `randomWordGenerator` will all have the same length during the execution of the program. Each time you call `randomWordGenerator` you will get a different word, but all of them will have the same length. If you re-run the program a second time, `n` will probably get a different value and the length of the words returned by `randomWordGenerator` will be different.
+*/
+let n: Int = .random(in: 0..<10)
+let randomWordGenerator: () -> String = { (0..<n).forEach { _ in .random(in: "a"..."z") } }
```

Comment:
Just formatting things a bit for easier readability.

```suggestion
let n: Int = .random(in: 0 ..< 10)
let randomWordGenerator: () -> String = { 
    (0 ..< n).forEach { _ in 
        .random(in: "a" ... "z") 
    }
}
```